rfcs/2020-09-29-4155-aws-s3-source.md

```diff
@@ -0,0 +1,253 @@
+  polling approach)
+
+## Drawbacks
+
+Maintenance burden of an additional source.
+
+## Alternatives
+
+There are a few alternate approaches ([some also noted by
+James](https://github.com/timberio/vector/issues/1017#issuecomment-699125610)).
+
+### SNS notifications
+
+AWS S3 can be configured to also send push-based notifications via SNS. This
+would require running Vector in a way where AWS can send requests to it.
+
+This may be a worthwhile approach in the future as it would result in faster
+object ingestion than polling SQS, but I think SQS's pull-based model is likely
+to be much more easy to use out-of-the-box as it does not require exposing
+a `vector` instance to incoming network requests and SQS consumers are more
+easily parallelizable (SNS would require a load balancer).
+
+I'd recommend tabling this one for now.
+
+### Polling
+
+As mentioned in
+[#1017](https://github.com/timberio/vector/issues/1017#issuecomment-694287111),
+another approach is to long-poll the bucket; listing all of the objects, and
+processing any new ones.
+
+This approach is liable to take a long time to process a bucket (hours or days
+with large buckets) and so is not the most generally applicable approach. It can
+work for small buckets or if a tightly scoped object prefix is used.
+
+This RFC will leave room for this, and possibly other, approaches to be added in
+the future.
+
+### Storing S3 object metadata in DynamoDB
+
+One recommendation in [a
+comment](https://github.com/timberio/vector/issues/1017#issuecomment-699125610)
+was to [store object created messages in
+DynamoDB](https://aws.amazon.com/blogs/big-data/building-and-maintaining-an-amazon-s3-metadata-index-without-servers/)
+(via Lambda). This could certainly work, but is a more complicated pipeline and
+I'm unsure how common it would be.
+
+I'd recommend tabling this one and see if anyone asks for it. I consider this
+approach representative of more complex S3 pipelines (e.g. you could imagine
+having the Lambda write the metadata to RDS instead).
+
+### Running Vector as a Lambda
+
+It is possible to have [AWS SQS directly call
+a Lambda](https://docs.aws.amazon.com/lambda/latest/dg/with-sqs.html) so a user
+could attempt to run Vector in a Lambda. We've generally tried to stay away from
+recommending running Vector in a Lambda due to the fact that Lambda's are
+largely stateless and some Vector components are stateful.
+
+## Outstanding Questions
+
+* We could attempt to make the compression handling more robust through falling
+  back to ["magic bytes"](https://en.wikipedia.org/wiki/List_of_file_signatures)
+  of the object to see if it _appears_ to be compressed.  We may want to have
+  the user opt into this behavior though as I could see it causing issues in
+  edge cases where the files are not compressed but happen to match the magic
+  bytes.
+
+## Plan Of Attack
+
+* [ ] Submit a PR with the initial implementation of source
+
+## Future work
```

Comment:
One other future item could be to integrate with some of the framing/codec work we've been talking about. This could help enable ingesting file formats that are not normal newline-delimited logs.